src/transform/src/dataflow.rs

```diff
@@ -434,76 +448,382 @@ pub fn optimize_dataflow_monotonic(dataflow: &mut DataflowDesc) -> Result<(), Tr
 }
 
 /// Restricts the indexes imported by `dataflow` to only the ones it needs.
+/// Additionally, it annotates each index import by how the index will be used, i.e., it fills in
+/// `IndexImport::usage_types`.
 ///
-/// The input `dataflow` should import all indexes belonging to all views it
+/// The input `dataflow` should import all indexes belonging to all views/sources/tables it
 /// references.
+///
+/// Should be called on a [DataflowDesc] only once.
 #[tracing::instrument(
     target = "optimizer",
     level = "debug",
     skip_all,
-    fields(path.segment ="index_imports")
+    fields(path.segment = "index_imports")
 )]
-fn optimize_dataflow_index_imports(
+fn prune_and_annotate_dataflow_index_imports(
     dataflow: &mut DataflowDesc,
     indexes: &dyn IndexOracle,
 ) -> Result<(), TransformError> {
-    // Generate (a mapping of views used by exports and objects to build) ->
-    // (indexes from that view that have been explicitly chosen to be used)
-    let mut indexes_by_view = BTreeMap::new();
-    for sink_desc in dataflow.sink_exports.iter() {
-        indexes_by_view
-            .entry(sink_desc.1.from)
-            .or_insert_with(BTreeSet::new);
+    // This will be a mapping of
+    // (ids used by exports and objects to build) ->
+    // (arrangement keys and usage types on that id that have been explicitly requested by the MIR
+    // plans)
+    let mut index_reqs_by_id = BTreeMap::new();
+
+    // First, insert `sink_exports` and `index_exports` into `index_reqs_by_id` to account for the
```

Comment:
I'm investigating this now, because it's confusing for me too.

> Got it, one of your comments below mentions that we need this to handle EXPLAIN CREATE INDEX.
However, it feels somewhat odd to me to render an item under the "used indexes" when it is in fact the index exported by the dataflow.

Actually, this is not for handling `EXPLAIN CREATE INDEX`. When a `PlanRoot` usage appears for a `CREATE INDEX` or `EXPLAIN CREATE INDEX`, it is because another, _imported_ index is used to build a new index directly (i.e., the user is probably just creating a new index on something that already has an index, but probably with a different key).

So this `sink_exports` and `index_exports` code is even more exotic:

`sink_exports`:
Turns out this is actually needed for a SUBSCRIBE to an indexed view, because this is handled by a dataflow that has an empty `objects_to_build`, and just imports an index, which it then exports as a sink:
```
subscribe to indexed view:

############ prune_and_annotate_dataflow_index_imports called with:
DataflowDescription {
	source_imports: {},
	index_imports: {User(3): IndexImport { desc: IndexDesc { on_id: User(2), key: [Column(0)] }, typ: RelationType { column_types: [ColumnType { scalar_type: Int32, nullable: false }], keys: [[]] }, monotonic: true, usage_types: None }},
	objects_to_build: [],
	index_exports: {},
	sink_exports: {Transient(2): ComputeSinkDesc { from: User(2), from_desc: RelationDesc { typ: RelationType { column_types: [ColumnType { scalar_type: Int32, nullable: false }], keys: [[]] }, names: [ColumnName("?column?")] }, connection: Subscribe(SubscribeSinkConnection), with_snapshot: true, up_to: Antichain { elements: [] } }},
	as_of: None, until: Antichain { elements: [] }, debug_name: "subscribe-t2"
}
```

In the other PR, I'm adding a special usage type for this called `SinkExport`. (And this PR was accounting this as a `FullScan`.)

`index_exports`:
I now think it can't happen at all that we are importing an index that we are also exporting, so I'll probably delete this code. (I'm running CI with Nightly now with an added assert that there is no intersection of index exports and imports.) I'll also ask QA to test that PR when it's ready (with lots of different top-level commands, e.g. SUBSCRIBE).

Reply:
Also note that `DataflowDescription::export_index` has the following comment:
"Future uses of `import_index` in other dataflow descriptions may use `id`,"
which also suggests that the same dataflow won't import it.